orangecontrib/text/preprocess.py

```diff
@@ -1,94 +1,161 @@
+                             "a list containing the stop words or None.")
+
+        # Transformation.
+        self.transformation = transformation
+        if transformation == 'porter_stemmer':
+            self.transformation = PorterStemmer
+        elif transformation == 'snowball_stemmer':
+            self.transformation = SnowballStemmer
+        elif transformation == 'lemmatizer':
+            self.transformation = Lemmatizer
+        elif transformation is not None:    # If not None and not supported.
+            raise ValueError("Transformation type '{}', is not recognized.".format(transformation))
 
     def __call__(self, data):
+        if isinstance(data, Corpus):
+            data = data.documents
+
         if isinstance(data, str):
-            vec = self.cv.fit([data])
-            return vec.get_feature_names()
-        if isinstance(data, list):
-            vec = self.cv.fit(data)
-            features = vec.get_feature_names()
-            docs = [[] for _ in range(len(data))]
-            for (line, column), count in vec.transform(data).todok().items():
-                docs[line].extend([features[column]] * count)
-            return docs
+            return self.preprocess_document(data)
+        elif isinstance(data, list):
+            return [self.preprocess_document(doc) for doc in data]
         else:
             raise ValueError("Type '{}' not supported.".format(type(data)))
 
+    def preprocess_document(self, data):
+        # Tokenize.
+        data = self.tokenize(data)
+        # Remove stop words.
+        if self.stop_words is not None:
+            data = self.remove_stop_words(data)
+        # Lowercase.
+        if self.lowercase:
+            data = [token.lower() for token in data]
+        # Transform.
+        if self.transformation is not None:
+            data = self.transform(data)
+        return data
+
+    def tokenize(self, data):
+        """
+        Splits the text of the input into tokens by whitespaces and punctuation.
+        :param data: The input holding the text to be tokenized. Can be provided
+            in form of a string or list of strings.
+        :type data: string, list
+        :return: list
+        """
+        if isinstance(data, str):
+            return self.tokenizer(data)
+        elif isinstance(data, Orange.data.Table):
+            # TODO: Handle this input.
+            return None
+        else:
+            raise ValueError("Cannot tokenize. Type {} not supported.".format(type(data)))
+
+    def transform(self, data):
+        """
+        Performs on the input, whatever transformation was specified at class instance creation.
+        :param data: The input holding the text to be transformed. Can be provided
+            in form of a string or list of string tokens.
+        :type data: string, list
+        :return: string, list
+        """
+        if self.transformation is None:
```

Comment:
This is checked twice, here and in the `preprocess_document` method. Since this method is only called from the `preprocess_document` and is not meant to be called directly, we should check this only once. Also, fix the same duplication in `remove_stop_words` method.